src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -851,39 +880,64 @@ int dhcp_server_handle_message(sd_dhcp_server *server, DHCPMessage *message,
 
                 /* verify that the requested address is from the pool, and either
                    owned by the current client or free */
-                if (pool_offset >= 0 &&
-                    server->bound_leases[pool_offset] == existing_lease) {
+                if (pool_offset >= 0 && static_lease) {
                         DHCPLease *lease;
```

Comment:
I don't think the `_cleanup_` macro should be used here, because if the server successfully sends an ACK message the pointer will be cleaned when it should be put in bound leases a bit further down:
```
server->bound_leases[pool_offset] = lease;
r = hashmap_put(server->leases_by_client_id, &lease->client_id, lease);
```
It's done the same way it was already done for existing leases.

Reply:
It is not explicitly documented, but we usually use `_cleanup_` macro where applicable, and use `TAKE_PTR()` macro on success branch, instead of `free()`ing memory in each failure branch.